roles/postgresql/tasks/main.yml

```diff
@@ -29,6 +21,76 @@
   register: "find_postgresql_data_path"
   tags: ["prepare", "prepare-postgresql"]
 
+- name: "Check if container {{ postgresql_container_name }} already exists on host"
+  community.docker.docker_container_info:
+    name: "{{ postgresql_container_name }}"
+  register: "postgresql_active_container"
+  tags: ["prepare", "prepare-postgresql"]
+
+- name: "Set postgresql_active_container_major_version"
+  ansible.builtin.set_fact:
+    postgresql_active_container_major_version: >-2
+      {{
+        postgresql_active_container.container.Config.Env
+        | select('search', '^PG_MAJOR=')
+        | first
+        | split('=')
+        | last
+      }}
+  when: "postgresql_active_container.exists"
+  tags: ["prepare", "prepare-postgresql"]
+
+- name: "Set postgresql_container_version to major version of existing container"
+  ansible.builtin.set_fact:
+    postgresql_container_version: "{{ postgresql_active_container_major_version }}"
+  when:
+    - "postgresql_active_container.exists"
+    - "not postgresql_container_custom_version"
+  tags: ["prepare", "prepare-postgresql"]
+
+- name: "Fail when installing postgresql_container_custom_version would require a database migration"
+  ansible.builtin.fail:
+    msg: |
+      postgresql_container_custom_version: {{ postgresql_container_custom_version }}
+      postgresql_active_container_major_version: {{ postgresql_active_container_major_version }}
+      postgresql_container_custom_version does not match the major version of the existing container.
+      PostgreSQL major version changes require a database migration, which this role does not provide.
+  when:
+    - "postgresql_active_container.exists"
+    - "postgresql_container_custom_version"
+    - "(postgresql_container_custom_version | int) != (postgresql_active_container_major_version | int)"
+  tags: ["prepare", "prepare-postgresql"]
+
+- name: "Set postgresql_container_version to postgresql_container_custom_version"
+  ansible.builtin.set_fact:
+    postgresql_container_version: "{{ postgresql_container_custom_version }}"
+  when: "postgresql_container_custom_version"
+  tags: ["prepare", "prepare-postgresql", "deploy", "deploy-postgresql"]
+
+- name: "Ensure image {{ postgresql_container_image_reference }} is pulled"
+  community.docker.docker_image:
+    name: "{{ postgresql_container_image_reference }}"
+    force_source: "{{ postgresql_container_pull }}"
+    source: "pull"
+    state: "present"
+  register: "postgresql_image_pulled"
+  tags: ["prepare", "prepare-postgresql"]
```

Comment:
I'm not 100% sure, but, I think this might only pull the image once and will never update it in the future.

What we could do, and what might help with the issue in the next task is having a version map like this:

```yaml
postgresql_versions:
  "16": "16.2"
  "15": "15.6"
  "14": "14.11"
  "13": "13.14"
```

Reply:
`community.docker.docker_image` seems to behave exactly like `docker pull`: providing only the major version always pulls the latest available minor version. At least according to my tests this works as intended.

I'm against mapping specific minor versions. IMHO the role should install the latest minor version or install the version specified by the user.

Reply:
If that works, then I'm happy with that part as is. I understood https://docs.ansible.com/ansible/latest/collections/community/docker/docker_image_module.html#parameter-force_source to mean that it'll only pull if there is no local image yet that satisfies what we want yet, which is how we're using it right now. If that assumption is wrong, we should talk about that again in the daily tomorrow, because our usage of the `docker_image` module at the moment is more convoluted than it needs to be then.

Reply:
I had a look at the module source code: https://github.com/ansible-collections/community.docker/blob/d4b654793eefc2d4e0fe52ca19462e82be834ee5/plugins/modules/docker_image.py#L512

I seems you were right: The image doesn't get pulled if one matching that name exists locally.

`postgresql_container_pull` defaults to `true`, that's why it worked during my tests...

Reply:
That should probably not default to true..

Reply:
I'm hesitant to change the default, but if desired we could set `postgresql_container_pull` to false if a `<major>.<minor>` version is given.